modules/gateway/peers.go

```diff
@@ -348,6 +403,45 @@ func acceptConnVersionHandshake(conn net.Conn, version string) (remoteVersion st
 	return remoteVersion, nil
 }
 
+// acceptConnSessionHandshake performs the session handshake and should be
+// called on the side accepting a connection request.
+// if err == nil, wantConn indicates if a connection is wanted on both sides.
+func acceptConnSessionHandshake(conn net.Conn, header sessionHeader) (wantConn bool, err error) {
+	var remoteHeader sessionHeader
+	// Read remote header.
+	if err = encoding.ReadObject(conn, &remoteHeader, build.MaxEncodedVersionLength); err != nil {
+		err = fmt.Errorf("failed to read remote header: %v", err)
+		return
+	}
+	// ensure that their genesisID is the same as ours
+	if bytes.Compare(header.GenesisID[:], remoteHeader.GenesisID[:]) != 0 {
+		header.WantConn = false
+		if err := encoding.WriteObject(conn, header); err != nil {
+			return false, fmt.Errorf("failed to write reject header: %v", err)
+		}
+		err = errPeerRejectedConn
+		return
+	}
+	// ensure that we're not connecting to ourselves
+	if bytes.Compare(header.UniqueID[:], remoteHeader.UniqueID[:]) == 0 {
+		header.WantConn = false
+		if err := encoding.WriteObject(conn, header); err != nil {
+			return false, fmt.Errorf("failed to reject header: %v", err)
+		}
+		err = errPeerExists
+		return
+	}
```

Comment:
define a helper function for this since it's used in two functions:
```go
// acceptableHeader returns an error if the remote header indicates that
// a connection is impossible.
func acceptableHeader(our, their sessionHeader) error {
    if their.GenesisID != our.GenesisID {
        return errPeerGenesisID
    } else if their.UniqueID == our.UniqueID {
        return errOurAddress
    } else if !their.WantConn {
        return errPeerRejected
    }
    return nil
}
```

Then `connectSessionHandshake` becomes:
```go
func connectSessionHandshake(conn net.Conn, header sessionHeader) error {
    // Write our header.
    if err := encoding.WriteObject(conn, header); err != nil {
        return fmt.Errorf("failed to write header: %v", err)
    }
    // Read remote header.
    var remoteHeader sessionHeader
    if err := encoding.ReadObject(conn, &remoteHeader, build.MaxEncodedVersionLength); err != nil {
        return fmt.Errorf("failed to read remote header: %v", err)
    }
    return acceptableHeader(header, remoteHeader)
}
```
And `acceptConnSessionHandshake` becomes:
```go
func acceptConnSessionHandshake(conn net.Conn, header sessionHeader) error {
    var remoteHeader sessionHeader
    // Read remote header.
    if err := encoding.ReadObject(conn, &remoteHeader, build.MaxEncodedVersionLength); err != nil {
	return fmt.Errorf("failed to read remote header: %v", err)
    }
    // Write our header.
    if err := encoding.WriteObject(conn, header); err != nil {
        return fmt.Errorf("failed to write header: %v", err)
    }
    return acceptableHeader(header, remoteHeader)
}
```